File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/abstracttype/ModExpr.java
package hu.bme.mit.theta.core.type.abstracttype;

import hu.bme.mit.theta.core.type.BinaryExpr;
import hu.bme.mit.theta.core.type.Expr;

import static hu.bme.mit.theta.core.utils.TypeUtils.cast;

public abstract class ModExpr<ExprType extends Divisible<ExprType>> extends BinaryExpr<ExprType, ExprType> {

    protected ModExpr(final Expr<ExprType> leftOp, final Expr<ExprType> rightOp) {
        super(leftOp, rightOp);
    }

    public static <ExprType extends Divisible<ExprType>> ModExpr<?> create2(final Expr<?> leftOp, final Expr<?> rightOp) {
        @SuppressWarnings("unchecked") final ExprType type = (ExprType) leftOp.getType();
        final Expr<ExprType> newLeftOp = cast(leftOp, type);
        final Expr<ExprType> newRightOp = cast(rightOp, type);
        return type.Mod(newLeftOp, newRightOp);
    }
}

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/abstracttype/ModExpr.kt
package hu.bme.mit.theta.core.type.abstracttype

import hu.bme.mit.theta.core.type.BinaryExpr
import hu.bme.mit.theta.core.type.Expr
import hu.bme.mit.theta.core.utils.TypeUtils.cast
import kotlinx.serialization.Serializable

/**
 * Abstract base class for modulo (mod) expressions over divisible types.
 */
@Serializable
abstract class ModExpr<ExprType : Divisible<ExprType>> : BinaryExpr<ExprType, ExprType>() {

    companion object {
        @JvmStatic
        fun <ExprType : Divisible<ExprType>> create2(
            leftOp: Expr<*>,
            rightOp: Expr<*>
        ): ModExpr<*> {
            @Suppress("UNCHECKED_CAST")
            val type = leftOp.type as ExprType
            val newLeftOp = cast(leftOp, type)
            val newRightOp = cast(rightOp, type)
            return type.Mod(newLeftOp, newRightOp)
        }
    }
}

